Core/include/Acts/TrackFinding/ComposableTrackStateCreator.hpp

```diff
@@ -0,0 +1,415 @@
+/// stateCandidateCreator
+/// @note if the number of states exceeds this number dynamic memory allocation will occur.
+///       the number is chosen to yield a container size of 64 bytes.
+static constexpr std::size_t s_maxBranchesPerSurface = 10;
+
+namespace CkfTypes {
+
+template <typename T>
+using BranchVector = boost::container::small_vector<T, s_maxBranchesPerSurface>;
+
+}  // namespace CkfTypes
+
+namespace detail {
+// helper method to deduce the return type and arguments of the member function
+template <auto memberfunc_ptr, typename return_t, typename instance_t,
+          typename... Args>
+inline auto makeDelegate(
+    [[maybe_unused]] return_t (instance_t::*memberfunc)(Args...) const,
+    const instance_t& instance) {
+  Delegate<return_t(Args...)> delegate;
+  delegate.template connect<memberfunc_ptr>(&instance);
+  return delegate;
+}
+// helper method to deduce the return type and arguments of the function
+template <auto func_ptr, typename return_t, typename... Args>
+inline auto makeDelegate([[maybe_unused]] return_t (*func)(Args...)) {
+  Delegate<return_t(Args...)> delegate;
+  delegate.template connect<func_ptr>();
+  return delegate;
+}
+}  // namespace detail
+
+/// @brief convenience method to create and connect a delegate.
+/// @tparam memberfunc_ptr a pointer to a member function of the class instance_t
+/// @param instance an instance which outlives this delegate.
+template <auto memberfunc_ptr, typename instance_t>
+inline auto makeDelegate(const instance_t& instance) {
+  return detail::makeDelegate<memberfunc_ptr>(memberfunc_ptr, instance);
+}
+
+/// @brief convenience method to create and connect a delegate.
+/// @tparam func_ptr a pointer to a function
+template <auto func_ptr>
+inline auto makeDelegate() {
+  return detail::makeDelegate<func_ptr>(func_ptr);
+}
+
+/// @brief Get source link range for the given surface and pass to derived class for
+///      the actual creation of track states.
+///
+/// - First get a source link range covering relevant measurements for the given
+/// surface
+///   This task is delegated to a SourceLinkAccessor.
+/// - Then, pass the source link range to the derived class which will do
+/// everything else.
+template <typename track_state_creator_derived_t,
+          typename source_link_iterator_t, typename track_container_t>
+struct ComposableTrackStateCreator {
+  using SourceLinkAccessor = SourceLinkAccessorDelegate<source_link_iterator_t>;
+  using TrackStatesResult =
+      Acts::Result<CkfTypes::BranchVector<TrackIndexType>>;
+  using TrackStateContainerBackend =
+      typename track_container_t::TrackStateContainerBackend;
+  using TrackProxy = typename track_container_t::TrackProxy;
+  using TrackStateProxy = typename track_container_t::TrackStateProxy;
+  using BoundState = std::tuple<BoundTrackParameters, BoundMatrix, double>;
+
+  SourceLinkAccessor sourceLinkAccessor;
+
+  const track_state_creator_derived_t& derived() const {
```

Comment:
This could be private maybe.

Reply:
Not sure whether that is something that needs to be hidden from the public, but it is presumably also not really necessary  to keep it public:

Reply:
IMO it is always a good idea to keep the public surface as small as possible. I don't think this is useful to the user?

Reply:
done